vicodyn/include/cocaine/vicodyn/peer.hpp

```diff
@@ -129,23 +140,31 @@ class peers_t {
     }
 
 
-    peers_t(context_t& context);
+    peers_t(context_t& context, const dynamic_t& args);
 
     auto register_peer(const std::string& uuid, const endpoints_t& endpoints, dynamic_t::object_t extra) -> std::shared_ptr<peer_t>;
 
     auto register_peer(const std::string& uuid, std::shared_ptr<peer_t> peer) -> void;
 
     auto erase_peer(const std::string& uuid) -> void;
 
-    auto register_app( const std::string& uuid, const std::string& name) -> void;
+    auto register_app(const std::string& uuid, const std::string& name) -> void;
 
     auto erase_app(const std::string& uuid, const std::string& name) -> void;
 
     auto ban_app(const std::string& uuid, const std::string& name, const std::chrono::milliseconds& timeout) -> void;
 
+    auto add_app_request_timing(const std::string& uuid, const std::string& name, clock_t::time_point start,
+                    clock_t::duration elapsed) -> void;
+
     auto erase(const std::string& uuid) -> void;
 
     auto peer(const std::string& uuid) -> std::shared_ptr<peer_t>;
+
+    auto choose_random(const std::string& app_name, peer_f peer_predicate, app_service_f app_service_predicate) const
```

Comment:
Both versions of `choose_random` seems to overlap a bit in main loop bodies and in peer choose part, is it possible to move common parts into utility functions/templates in order to avoid code repetition?

Reply:
Ok, I'll think about it more.

Reply:
Fixed.